File: relisoft-hr/server/controllers/visitorController.js
import Visitor from '../models/Visitor.js';
import Visit from '../models/Visit.js';
import Employee from '../models/Employee.js';

export const registerVisitor = async (req, res) => {
  try {
    const visitor = await Visitor.create(req.body);
    res.status(201).json({ success: true, data: visitor });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const getVisitors = async (req, res) => {
  try {
    const { isBlacklisted } = req.query;
    const query = {};
    if (isBlacklisted !== undefined) query.isBlacklisted = isBlacklisted === 'true';

    const visitors = await Visitor.find(query).sort('-createdAt');
    res.status(200).json({ success: true, data: visitors });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const updateVisitor = async (req, res) => {
  try {
    const visitor = await Visitor.findByIdAndUpdate(req.params.id, req.body, { new: true, runValidators: true });
    if (!visitor) return res.status(404).json({ success: false, message: 'Visitor not found' });
    res.status(200).json({ success: true, data: visitor });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const scheduleVisit = async (req, res) => {
  try {
    const employee = await Employee.findOne({ userId: req.user.id });
    const visit = await Visit.create({ ...req.body, host: employee?._id || req.body.host });
    res.status(201).json({ success: true, data: visit });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const getPendingVisits = async (req, res) => {
  try {
    const { host } = req.query;
    const query = { status: 'Pending' };
    if (host) query.host = host;

    const visits = await Visit.find(query)
      .populate('visitor', 'firstName lastName company email phone')
      .populate('host', 'firstName lastName')
      .sort('-expectedDate');
    res.status(200).json({ success: true, data: visits });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const approveVisit = async (req, res) => {
  try {
    const { approved, notes } = req.body;
    const employee = await Employee.findOne({ userId: req.user.id });

    const update = {
      status: approved ? 'Approved' : 'Cancelled',
      approvedBy: employee?._id,
      notes,
    };

    if (approved) {
      const gatePassNumber = `VP-${String(Math.floor(Math.random() * 100000)).padStart(5, '0')}`;
      update.gatePassNumber = gatePassNumber;
    }

    const visit = await Visit.findByIdAndUpdate(req.params.id, update, { new: true });
    if (!visit) return res.status(404).json({ success: false, message: 'Visit not found' });
    res.status(200).json({ success: true, data: visit });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const checkIn = async (req, res) => {
  try {
    const visit = await Visit.findByIdAndUpdate(
      req.params.id,
      { status: 'CheckedIn', checkedInAt: Date.now() },
      { new: true }
    );
    if (!visit) return res.status(404).json({ success: false, message: 'Visit not found' });
    res.status(200).json({ success: true, data: visit });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const checkOut = async (req, res) => {
  try {
    const visit = await Visit.findByIdAndUpdate(
      req.params.id,
      { status: 'CheckedOut', checkedOutAt: Date.now() },
      { new: true }
    );
    if (!visit) return res.status(404).json({ success: false, message: 'Visit not found' });
    res.status(200).json({ success: true, data: visit });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const getVisitHistory = async (req, res) => {
  try {
    const { visitorId } = req.params;
    const visits = await Visit.find({ visitor: visitorId })
      .populate('visitor', 'firstName lastName company')
      .populate('host', 'firstName lastName')
      .populate('approvedBy', 'firstName lastName')
      .sort('-expectedDate');
    res.status(200).json({ success: true, data: visits });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

File: relisoft-hr/server/Controllers/VisitorController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using RelisoftHR.Data;
using RelisoftHR.Models;
using RelisoftHR.Services;
 
namespace RelisoftHR.Controllers;

[ApiController]
[Route("api/visitors")]
[Authorize]
public class VisitorController : ControllerBase
{
    private readonly AppDbContext _db;
    private readonly NotificationHelper _notif;
    public VisitorController(AppDbContext db, NotificationHelper notif)
    {
        _db = db;
        _notif = notif;
    }

    private int GetUserId()
    {
        var claim = User.FindFirst(System.Security.Claims.ClaimTypes.NameIdentifier)?.Value;
        return claim != null ? int.Parse(claim) : 0;
    }

    [HttpGet]
    public async Task<ActionResult> GetVisitors([FromQuery] string? status, [FromQuery] string? date)
    {
        var query = _db.Visitors.AsQueryable();
        if (!string.IsNullOrEmpty(status))
            query = query.Where(v => v.Status == status);
        if (!string.IsNullOrEmpty(date) && DateTime.TryParse(date, out var d))
            query = query.Where(v => v.ExpectedDate.Date == d.Date);
        var list = await query.OrderByDescending(v => v.ExpectedDate).ThenByDescending(v => v.ExpectedTime).ToListAsync();
        return Ok(list);
    }

    [HttpPost]
    public async Task<ActionResult> CreateVisitor([FromBody] Visitor req)
    {
        req.CreatedOn = DateTime.UtcNow;
        _db.Visitors.Add(req);
        await _db.SaveChangesAsync();

        if (req.HostEmployeeId.HasValue)
        {
            var host = await _db.Employees.FindAsync(req.HostEmployeeId.Value);
            if (host != null)
            {
                await _notif.NotifyEmployeeAsync(host.Id, host, "Visitor Registered",
                    $"{req.FullName} has been registered to visit you.", "visitor",
                    "Visitor Registered", EmailTemplates.VisitorRegistered(host.FullName, req.FullName, req.ExpectedDate),
                    link: "/visitors");
            }
        }

        return Ok(req);
    }

    [HttpPut("{id}")]
    public async Task<ActionResult> UpdateVisitor(int id, [FromBody] Visitor req)
    {
        var v = await _db.Visitors.FindAsync(id);
        if (v == null) return NotFound();
        v.FullName = req.FullName;
        v.Email = req.Email;
        v.Phone = req.Phone;
        v.Company = req.Company;
        v.VisitingEmployee = req.VisitingEmployee;
        v.Purpose = req.Purpose;
        v.ExpectedDate = req.ExpectedDate;
        v.ExpectedTime = req.ExpectedTime;
        v.Notes = req.Notes;
        v.HostEmployeeId = req.HostEmployeeId;
        await _db.SaveChangesAsync();
        return Ok(v);
    }

    [HttpPost("{id}/check-in")]
    public async Task<ActionResult> CheckIn(int id)
    {
        var v = await _db.Visitors.FindAsync(id);
        if (v == null) return NotFound();
        v.CheckInTime = DateTime.UtcNow;
        v.Status = "CheckedIn";
        await _db.SaveChangesAsync();
        return Ok(v);
    }

    [HttpPost("{id}/check-out")]
    public async Task<ActionResult> CheckOut(int id)
    {
        var v = await _db.Visitors.FindAsync(id);
        if (v == null) return NotFound();
        v.CheckOutTime = DateTime.UtcNow;
        v.Status = "CheckedOut";
        await _db.SaveChangesAsync();
        return Ok(v);
    }

    [HttpGet("today")]
    public async Task<ActionResult> GetTodayVisitors()
    {
        var today = DateTime.UtcNow.Date;
        var list = await _db.Visitors
            .Where(v => v.ExpectedDate.Date == today)
            .OrderBy(v => v.ExpectedTime)
            .ToListAsync();
        return Ok(list);
    }
}
